domino/datasets.py

```diff
@@ -0,0 +1,190 @@
+            routes: _Routes,
+            target_relative_path: str,
+
+            file_upload_setting: str,
+            max_workers: int,
+            target_chunk_size: int,
+
+    ):
+        self.csrf_no_check_header = csrf_no_check_header
+        self.dataset_id = dataset_id
+        self.local_path_file_or_directory = local_path_to_file_or_directory
+        self.log = log
+        self.request_manager = request_manager
+        self.routes = routes
+        self.target_chunk_size = target_chunk_size or 8 * MB
+        self.file_upload_setting = file_upload_setting or FILE_UPLOAD_SETTING_DEFAULT
+        self.max_workers = max_workers or MAX_WORKERS
+        self.target_relative_path = target_relative_path
+        self.upload_key = None  # this will be set once the session is started
+
+    def start_upload_session(self):
+        start_upload_body = {
+            "filePaths": [],
+            "fileCollisionSetting": self.file_upload_setting
+        }
+        start_upload_url = self.routes.datasets_start_upload(self.dataset_id)
+        self.upload_key = self.request_manager.post(start_upload_url, json=start_upload_body).json()
+        if not self.upload_key:
+            raise RuntimeError(f"upload key for {self.dataset_id} not found. Session could not start.")
+
+    def upload(self):
+        if not self.upload_key:
+            raise RuntimeError(f"upload key for {self.dataset_id} not found. Please start session before uploading.")
+        q = self._create_chunk_queue()
+        with ThreadPoolExecutor(self.max_workers) as executor:
+            # list ensures all the threads are complete before returning results
+            results = list(executor.map(self._upload_chunk, q))
+        return results
+
+    def cancel_upload_session(self):
+        url = self.routes.datasets_cancel_upload(self.dataset_id, self.upload_key)
+        return self.request_manager.get(url)
+
+    def end_upload_session(self):
+        if not self.upload_key:
+            raise RuntimeError(f"upload key for {self.dataset_id} not found. Could not end session.")
+        url = self.routes.datasets_end_upload(self.dataset_id, self.upload_key, self.target_relative_path)
+        return self.request_manager.get(url)
+
+    def _create_chunk_queue(self) -> [UploadChunk]:
+        if os.path.isfile(self.local_path_file_or_directory):
+            return self._create_chunk(self.local_path_file_or_directory)
+        else:
+            nested_q = []
+            for dirpath, dirnames, filenames in os.walk(self.local_path_file_or_directory):
+                for filename in filenames:
+                    # construct the relative path for each file
+                    relative_path_to_file = os.path.join(dirpath, filename)
+                    # append chunk to queue
+                    nested_q.append(self._create_chunk(relative_path_to_file))
+            # flattening list of chunks
+            return [chunk for chunks in nested_q for chunk in chunks]
```

Comment:
WAT is happening here 😆 . Can you name things differently? is there a way to flatten this in the previous for loop?